Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+
+lemma coverEntropySupUni_antitone (T : X → X) (F : Set X) :
+    Antitone (fun U : Set (X × X) ↦ coverEntropySupUni T F U) :=
+  fun _ _ U_V ↦ (limsup_le_limsup) <| Eventually.of_forall
+    fun n ↦ monotone_div_right_of_nonneg (Nat.cast_nonneg' n)
+    <| log_monotone (ENat.toENNReal_mono (coverMincard_antitone_entourage T F n U_V))
+
+lemma coverEntropyInfUni_le_coverEntropySupUni (T : X → X) (F : Set X) (U : Set (X × X)) :
+    coverEntropyInfUni T F U ≤ coverEntropySupUni T F U := liminf_le_limsup
+
+@[simp]
+lemma coverEntropySupUni_empty {T : X → X} {U : Set (X × X)} :
+    coverEntropySupUni T ∅ U = ⊥ := by
+  suffices h : ∀ᶠ n : ℕ in atTop, log (coverMincard T ∅ U n) / n = ⊥ by
+    rw [coverEntropySupUni]
+    exact limsup_congr h ▸ limsup_const ⊥
+  · simp only [coverMincard_empty, ENat.toENNReal_zero, log_zero, eventually_atTop]
+    exact ⟨1, fun n n_pos ↦ bot_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)⟩
+
+@[simp]
+lemma coverEntropyInfUni_empty {T : X → X} {U : Set (X × X)} :
+    coverEntropyInfUni T ∅ U = ⊥ :=
+  eq_bot_mono (coverEntropyInfUni_le_coverEntropySupUni T ∅ U) coverEntropySupUni_empty
+
+lemma coverEntropyInfUni_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ coverEntropyInfUni T F U :=
+  (le_iInf fun n ↦ div_nonneg (log_coverMincard_nonneg T h U n) (Nat.cast_nonneg' n)).trans
+    iInf_le_liminf
+
+lemma coverEntropySupUni_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) :
+    0 ≤ coverEntropySupUni T F U :=
+  (coverEntropyInfUni_nonneg T h U).trans (coverEntropyInfUni_le_coverEntropySupUni T F U)
+
+lemma coverEntropyInfUni_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    coverEntropyInfUni T F univ = 0 := by
+  simp [coverEntropyInfUni, coverMincard_univ T h]
+
+lemma coverEntropySupUni_univ (T : X → X) {F : Set X} (h : F.Nonempty) :
+    coverEntropySupUni T F univ = 0 := by
+  simp [coverEntropySupUni, coverMincard_univ T h]
+
+lemma coverEntropyInfUni_le_log_coverMincard_div {T : X → X} {F : Set X} (F_inv : MapsTo T F F)
+    {U : Set (X × X)} (U_symm : SymmetricRel U) {n : ℕ} (n_pos : 0 < n) :
+    coverEntropyInfUni T F (U ○ U) ≤ log (coverMincard T F U n) / n := by
+  apply liminf_le_of_frequently_le'
+  rw [frequently_atTop]
+  refine fun N ↦ ⟨(max 1 N) * n, ?_, ?_⟩
+  · rcases eq_zero_or_pos N with rfl | N_pos
+    · exact zero_le ((max 1 0) * n)
+    · rw [max_eq_right (Nat.one_le_of_lt N_pos)]
+      nth_rw 2 [← mul_one N]
+      exact Nat.mul_le_mul_left N (Nat.one_le_of_lt n_pos)
+  · have := log_coverMincard_iterate_le F_inv U_symm n
+      (lt_of_lt_of_le zero_lt_one (le_max_left 1 N))
+    rw [mul_comm n (max 1 N)] at this
+    apply le_of_eq_of_le _ (div_le_div_right_of_nonneg (Nat.cast_nonneg' n) this)
+    rw [div_div]
+    congr
+    exact natCast_mul (max 1 N) n
+
+lemma coverEntropyInfUni_le_log_card_div {T : X → X} {F : Set X} (F_inv : MapsTo T F F)
```

Comment:
Maybe put this in the `IsDynCoverOf` namespace? Also, wouldn't it be better to give it for the `SupUni`?

Reply:
Yes and yes. Done.